BUG: 2D ndarray of dtype 'object' is always copied upon construction

irgolic opened the pull request:
This change is aesthetically unpleasing, it indents too far. #26825 already suggested to separate this part into a function. Please let me know what exactly I should abstract into a function, and where to put it. Alternatively, amend this PR yourself.

Thanks for pandas. :)

- [X] closes #39263
- [X] tests added / passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [X] whatsnew entry

jreback commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -231,6 +231,7 @@ Datetimelike
 - Bug in :meth:`DatetimeIndex.intersection`, :meth:`DatetimeIndex.symmetric_difference`, :meth:`PeriodIndex.intersection`, :meth:`PeriodIndex.symmetric_difference` always returning object-dtype when operating with :class:`CategoricalIndex` (:issue:`38741`)
 - Bug in :meth:`Series.where` incorrectly casting ``datetime64`` values to ``int64`` (:issue:`37682`)
 - Bug in :class:`Categorical` incorrectly typecasting ``datetime`` object to ``Timestamp`` (:issue:`38878`)
+- Bug in :func:`pandas.core.internals.construction.init_ndarray` unnecessarily copying all object arrays after datetime inference
```
we don't add a whatsnew for internal functions, rather you want to note what a user would care about. so you can say about the same thing, but its about the DataFrame constructor.

jreback replied:
add the issue number

jreback commented on pandas/core/internals/construction.py:
```diff
@@ -240,18 +241,35 @@ def init_ndarray(values, index, columns, dtype: Optional[DtypeObj], copy: bool):
-                make_block(dvals_list[n], placement=[n], ndim=2)
-                for n in range(len(dvals_list))
-            ]
+            i = 0
```
yeah this is not great. we *already* do this type of dtype grouping *inside* the block manager constructors. much better to do there. However this might be non-trivial and/or might have a bit impact performance.

so i would try to re-use one of the existing grouping rountines here instead.

irgolic replied:
Could you point me to the module?

jbrockmendel replied:
i dont think we have anything quite like this, in part because i wrote something similar in #33597

jbrockmendel replied:
what if you dont do the make_block calls here, but instead keep construct the arrays, and then instead of calling create_block_manager_from_blocks at the end could call create_block_manager_from_arrays?

irgolic replied:
That solves it but has a ripple effect of making other stuff fail. That constructor might need to be adjusted for use in instantiating a DataFrame.
![Screenshot 2021-01-20 at 02 14 22](https://user-images.githubusercontent.com/24586651/105117965-5bcab300-5ac5-11eb-8449-ee574c0696b3.png)

jbrockmendel replied:
can you post one of the actual tracebacks?

irgolic replied:
Some of these were a missing ValueError being raised in `_form_blocks`, but it still ended up copying everything – that's just how `create_block_manager_from_arrays` is implemented. I've implemented a new `_from_array` constructor, which will falls back to `_form_blocks` in the case of a datetime conversion, but otherwise doesn't copy.

irgolic replied:
I tried to avoid adding a kwarg to the constructor, but it seemed like that would get messier.

jreback reviewed (changes requested):
you will have to measure performance if it changes in any way. run the asv suite. and see which benchmarks are affected.

I am not sure how this change actually helps anything. e.g. how does sharing memory actually help here? (because the second you do almost anything these blocks will likely get split apart and copied again). 

so having a specific benchmark is really important here. sharing memory is not that interesting as its fleeting.

jreback commented:
cc @jbrockmendel

jbrockmendel commented:
side note, after #38939 we can also avoid consolidating in this case, which would avoid making copies

jbrockmendel commented on pandas/core/internals/construction.py:
```diff
@@ -240,18 +241,35 @@ def init_ndarray(values, index, columns, dtype: Optional[DtypeObj], copy: bool):
+                    )
+                else:
+                    block_values.extend(
+                        make_block(
```
what happens here if we have two consecutive datetime64tz columns?

irgolic replied:
```python
In[5]: block_values
Out[5]: 
[DatetimeTZBlock: slice(0, 1, 1), 1 x 1, dtype: datetime64[ns, UTC],
 DatetimeTZBlock: slice(1, 2, 1), 1 x 1, dtype: datetime64[ns, UTC],
 DatetimeTZBlock: slice(2, 3, 1), 1 x 1, dtype: datetime64[ns, UTC]]
```

irgolic commented:
From what I can gather, the problem is that not all datetimelike types are `dtype`s. Checking equality between a `dtype` and a non-`dtype` throws an exception, which is likely why this wasn't implemented in the block manager constructor (where I assume everything works in `dtype`s?)

jbrockmendel commented:
> Checking equality between a dtype and a non-dtype throws an exception

Im guessing the issue is trying to do `some_numpy_dtype == some_extension_dtype` which does raise.  if you use is_dtype_equal(some_numpy_dtype, some_extension_dtype)` you'll get a bool back

pep8speaks commented:
Hello @irgolic! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/core/internals/managers.py`](https://github.com/pandas-dev/pandas/blob/322a677c0aecb2c23b0517ffc614563e4b5876a9/pandas/core/internals/managers.py):

> [Line 1696:89](https://github.com/pandas-dev/pandas/blob/322a677c0aecb2c23b0517ffc614563e4b5876a9/pandas/core/internals/managers.py#L1696): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 88 characters)

irgolic commented:
> I am not sure how this change actually helps anything. e.g. how does sharing memory actually help here? (because the second you do almost anything these blocks will likely get split apart and copied again).

I'm a core developer of https://github.com/biolab/orange3, focusing on UX. Orange's data objects are also based on numpy arrays, Orange's `Table` exposes them as `table.X`, `table.Y`, `table.metas`. I'm trying to wrap our numpy arrays in DataFrame objects, and allow people to use pandas as an accessor in Table (`table.X_df`, `table.Y_df`, `table.metas_df`) (https://github.com/biolab/orange3/pull/5189).

While you're right in that most operations will split it apart anyway, in my use case, it'll avoid an unnecessary copy operation.

> you will have to measure performance if it changes in any way. run the asv suite. and see which benchmarks are affected.

On it. If none of the benchmarks change, I'll add one.

jreback commented on pandas/core/internals/managers.py:
```diff
@@ -1679,6 +1681,41 @@ def create_block_manager_from_arrays(
         raise construction_error(len(arrays), arrays[0].shape, axes, e)
 
 
+def create_block_manager_from_array(
```
can you not *just* call create_block_manager_from_arrays? and if not why not?

irgolic replied:
Because `create_block_manager_from_arrays` (and `_form_blocks`, and `_stack_arrays`) is written to receive a list of 1-dimensional columns, and create copy of them. My latest commit adjusts these functions to avoid copying when necessary.

jreback commented on pandas/core/internals/managers.py:
```diff
@@ -1705,26 +1706,42 @@ def construction_error(tot_items, block_shape, axes, e=None):
-    names_idx = names
-    if names_idx.equals(axes[0]):
-        names_indexer = np.arange(len(names_idx))
+    if len(arrays) == 1:
```
for sure needs comments, and even better pls separate this out to a new function.

irgolic replied:
Better?

jreback replied:
no, this looks almost like a groupby to me. i am puzzled why the axes are involved here at all.

irgolic replied:
That's how it was already written. It finds all variables of the same block type, and makes blocks out of them. If you pass in more arrays than names, only those arrays whose names were specified are blockified.

I could rewrite this into a groupby, I guess. It'll create more contiguous blocks than it does currently.

If you look at `create_block_manager_from_blocks`, you'll see that it allows passing a singular array (not block), and it'll create a contiguous block for it. I could do this in `create_block_manager_from_arrays`, but IIRC that made a lot of things fail, like `init_dict`, which uses the same constructor. Maybe #38939 will fix that.

jreback commented on pandas/core/internals/managers.py:
```diff
@@ -1836,16 +1855,24 @@ def _shape_compat(x) -> Shape:
     first = arrays[0]
-    shape = (len(arrays),) + _shape_compat(first)
 
+    if len(arrays) == 1:
```
add omments

jreback commented:
I think this PR will be almost trivial once https://github.com/pandas-dev/pandas/pull/38939 is merged

irgolic commented:
I've added a comment to that PR. It may or may not be trivial depending on the answer.
https://github.com/pandas-dev/pandas/pull/38939/files#r562162332

irgolic commented:
This is necessary because of the way consolidation works, specifically the nature of `pandas.core.internals.managers._merge_blocks`:
```python
def _merge_blocks(
    blocks: List[Block], dtype: DtypeObj, can_consolidate: bool
) -> List[Block]:

    if len(blocks) == 1:
        return blocks

    if can_consolidate:

        if dtype is None:
            if len({b.dtype for b in blocks}) != 1:
                raise AssertionError("_merge_blocks are invalid!")

        # TODO: optimization potential in case all mgrs contain slices and
        # combination of those slices is a slice, too.
        new_mgr_locs = np.concatenate([b.mgr_locs.as_array for b in blocks])
        new_values = np.vstack([b.values for b in blocks])

        argsort = np.argsort(new_mgr_locs)
        new_values = new_values[argsort]
        new_mgr_locs = new_mgr_locs[argsort]

        return [make_block(new_values, placement=new_mgr_locs, ndim=2)]

    # can't consolidate --> no merge
    return blocks
```
Meaning, if you create a block for each column of the array, it won't understand that these are all slices of the same contiguous thing. To my knowledge, there's no easy way to check this. So we make sure to form one block for contiguous data.

irgolic committed:
BUG: 2D ndarray of dtype 'object' is always copied upon construction (#39263)

irgolic commented:
You were right @jreback, the change can be much simpler than I wrote previously. If were not to try to jam everything into `create_block_manager_from_arrays`, it works out quite nicely.

Still, it'd be much nicer to make a `create_block_manager_from_array` constructor. `create_block_manager_from_blocks` is now only called with a single array, which is an ugly way of doing this:
```python
def create_block_manager_from_array(
    array,
    axes: List[Index],
    consolidate: bool = True,
) -> BlockManager:
    try:
        block = make_block(values=array, placement=slice(0, len(axes[0])), ndim=2)
        mgr = BlockManager([block], axes)
    except ValueError as e:
        raise construction_error(array.shape[0], array.shape[1:], axes, e)
    if consolidate:
        mgr._consolidate_inplace()
    return mgr
```

If you agree, I can add this after #38939 is merged.

jreback commented on pandas/core/internals/managers.py:
```diff
@@ -1634,21 +1634,17 @@ def fast_xs(self, loc):
-                        values=blocks[0], placement=slice(0, len(axes[0])), ndim=2
-                    )
-                ]
+    if len(blocks) == 1 and not isinstance(blocks[0], Block):
```
didn't you have a suggestion to simplify this? 
wonder where we actually hit this (the no block case)

irgolic replied:
Whoops, the test got lost as I was fiddling with local branches. Gimme a sec.

I saw that `create_block_manager_from_blocks` was left in the API for backward compatibility, I didn't want to fiddle with it.

jreback replied:
if you can remove / simplify happy to have that

jreback reviewed (changes requested):
didn't you have a test for this?

jreback commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -237,6 +237,7 @@ Datetimelike
 - Bug in comparisons between :class:`Timestamp` object and ``datetime64`` objects just outside the implementation bounds for nanosecond ``datetime64`` (:issue:`39221`)
 - Bug in :meth:`Timestamp.round`, :meth:`Timestamp.floor`, :meth:`Timestamp.ceil` for values near the implementation bounds of :class:`Timestamp` (:issue:`39244`)
 - Bug in :func:`date_range` incorrectly creating :class:`DatetimeIndex` containing ``NaT`` instead of raising ``OutOfBoundsDatetime`` in corner cases (:issue:`24124`)
+- Bug in :class:`DataFrame` constructor always copying 2D object arrays
```
add this PR number as the issue.

jreback commented on pandas/core/internals/construction.py:
```diff
@@ -236,30 +236,26 @@ def init_ndarray(values, index, columns, dtype: Optional[DtypeObj], copy: bool):
+            array = maybe_infer_to_datetimelike(values)
     else:
-        block_values = [values]
+        array = values
```
can you assign `array=values`  before L238 and remove the appropriate else clauses

jreback commented:
latest looks very nice. let's hope everything passes :->

irgolic commented:
:D

jreback commented:
@irgolic can you run the construction asv's and report on the results here. to see if anything interesting. also the indexing might be useful to run (note that can take a while to run :-<)

cc @jbrockmendel if comments

irgolic commented:
Construction asv says `BENCHMARKS NOT SIGNIFICANTLY CHANGED.`

I'll post indexing results once they're done.

irgolic commented:
```
       before           after         ratio
     [7e531e3d]       [7157e6aa]
                      <2d-object-dont-copy>
+        67.8±1μs       98.5±0.9ms  1454.32  indexing.NumericSeriesIndexing.time_getitem_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+     6.34±0.03ms       7.06±0.2ms     1.11  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Int64Index'>, 'unique_monotonic_inc')
-        99.8±2ms      3.66±0.04ms     0.04  indexing.NumericSeriesIndexing.time_loc_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-        99.3±2ms          479±8μs     0.00  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-        99.2±2ms          171±2μs     0.00  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

I'm afraid I don't have enough knowledge to interpret these, but interesting how some rose and some dropped.

jreback commented:
can you repeat those benchamrks to see if they repro (the ones that you showed above)

irgolic commented:
```
       before           after         ratio
     [caf81fac]       [7157e6aa]
                      <2d-object-dont-copy>
+     4.64±0.06ms          326±4ms    70.28  indexing.NumericSeriesIndexing.time_loc_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+     5.07±0.06ms          327±4ms    64.58  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+     12.3±0.09ms          327±4ms    26.64  indexing.NumericSeriesIndexing.time_loc_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+      12.6±0.5ms          329±4ms    26.03  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+      37.0±0.3ms          332±4ms     8.96  indexing.NumericSeriesIndexing.time_getitem_lists(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
-       593±200μs         527±10μs     0.89  indexing.NonNumericSeriesIndexing.time_getitem_list_like('period', 'nonunique_monotonic_inc')
-       78.2±30μs         68.3±1μs     0.87  indexing.NonNumericSeriesIndexing.time_getitem_pos_slice('period', 'nonunique_monotonic_inc')
-      5.99±0.5μs      5.19±0.05μs     0.87  indexing.DataFrameStringIndexing.time_getitem_scalar
-     3.99±0.07μs      3.44±0.08μs     0.86  indexing.NonNumericSeriesIndexing.time_getitem_scalar('string', 'non_monotonic')
-        4.18±2ms      3.34±0.02ms     0.80  indexing.InsertColumns.time_assign_list_like_with_setitem
-       745±200μs         543±10μs     0.73  indexing.NonNumericSeriesIndexing.time_getitem_list_like('datetime', 'non_monotonic')
-       585±200μs          410±2μs     0.70  indexing.AssignTimeseriesIndex.time_frame_assign_timeseries_index
-       74.1±20ms      4.16±0.09ms     0.06  indexing.NonNumericSeriesIndexing.time_getitem_list_like('period', 'unique_monotonic_inc')
-       89.8±20ms       4.50±0.2ms     0.05  indexing.NonNumericSeriesIndexing.time_getitem_list_like('period', 'non_monotonic')
-      99.6±0.3ms      3.72±0.08ms     0.04  indexing.NumericSeriesIndexing.time_loc_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-      99.7±0.7ms         491±10μs     0.00  indexing.NumericSeriesIndexing.time_getitem_slice(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-      98.8±0.4ms          169±3μs     0.00  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

Was something related to indexing merged into master recently? Or are they generally a bit unreliable? I'm running these on a dedicated machine.

jreback commented:
you shouldn't have this kind of variation

irgolic commented:
Three more runs:

```
       before           after         ratio
     [caf81fac]       [7157e6aa]
                      <2d-object-dont-copy>
+        72.2±2μs       98.9±0.5ms  1369.02  indexing.NumericSeriesIndexing.time_getitem_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+         163±1μs       99.3±0.5ms   607.68  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+         469±6μs       99.2±0.4ms   211.69  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+         476±6μs       99.2±0.5ms   208.10  indexing.NumericSeriesIndexing.time_getitem_slice(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+     4.62±0.04ms          327±3ms    70.80  indexing.NumericSeriesIndexing.time_loc_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+     4.96±0.06ms          327±3ms    65.98  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+      12.1±0.1ms          329±3ms    27.23  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+     3.66±0.04ms       99.7±0.4ms    27.21  indexing.NumericSeriesIndexing.time_loc_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+     12.2±0.07ms          329±3ms    26.98  indexing.NumericSeriesIndexing.time_loc_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+      4.01±0.1ms         99.8±2ms    24.85  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+      36.2±0.4ms          331±3ms     9.15  indexing.NumericSeriesIndexing.time_getitem_lists(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+      5.96±0.2ms       6.61±0.3ms     1.11  indexing.NumericSeriesIndexing.time_loc_array(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'unique_monotonic_inc')
+      6.38±0.3ms       7.02±0.2ms     1.10  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Int64Index'>, 'unique_monotonic_inc')
-        494±90μs        255±0.1μs     0.52  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Int8Engine'>, <class 'numpy.int8'>), 'monotonic_incr')
```
```
       before           after         ratio
     [caf81fac]       [7157e6aa]
                      <2d-object-dont-copy>
+     4.77±0.04ms          322±4ms    67.53  indexing.NumericSeriesIndexing.time_loc_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+     5.02±0.04ms          325±3ms    64.61  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+      12.1±0.1ms          325±4ms    26.75  indexing.NumericSeriesIndexing.time_loc_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+      12.5±0.5ms          327±4ms    26.05  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+      36.8±0.3ms          329±3ms     8.96  indexing.NumericSeriesIndexing.time_getitem_lists(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
-     4.37±0.07ms      3.92±0.03ms     0.90  indexing.NonNumericSeriesIndexing.time_getitem_list_like('period', 'unique_monotonic_inc')
-      34.7±0.5μs       30.5±0.7μs     0.88  indexing.NonNumericSeriesIndexing.time_getitem_pos_slice('string', 'non_monotonic')
-        99.5±2ms      4.06±0.08ms     0.04  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
```
```
       before           after         ratio
     [caf81fac]       [7157e6aa]
                      <2d-object-dont-copy>
+         167±2μs       98.7±0.5ms   590.65  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+         469±8μs         99.2±2ms   211.53  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+     4.86±0.09ms          326±3ms    67.17  indexing.NumericSeriesIndexing.time_loc_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+     5.53±0.05ms          327±3ms    59.09  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+     12.3±0.09ms          329±4ms    26.71  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+     12.3±0.09ms          327±3ms    26.68  indexing.NumericSeriesIndexing.time_loc_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+      37.0±0.2ms          331±3ms     8.94  indexing.NumericSeriesIndexing.time_getitem_lists(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
-      10.4±0.9μs      9.42±0.07μs     0.90  indexing.CategoricalIndexIndexing.time_getitem_scalar('non_monotonic')
-      4.89±0.2ms       4.38±0.2ms     0.90  indexing.NonNumericSeriesIndexing.time_getitem_list_like('period', 'non_monotonic')
-     4.47±0.03ms      3.99±0.02ms     0.89  indexing.NonNumericSeriesIndexing.time_getitem_list_like('period', 'unique_monotonic_inc')
-         395±1μs         264±10μs     0.67  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Int8Engine'>, <class 'numpy.int8'>), 'monotonic_incr')
-        99.5±1ms      3.78±0.08ms     0.04  indexing.NumericSeriesIndexing.time_loc_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-        98.6±2ms       67.5±0.9μs     0.00  indexing.NumericSeriesIndexing.time_getitem_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
```
No idea where this variation comes from.

jreback commented:
these indicate your change is impacting

jreback commented:
my guess is that somewhere this is causing a numeric index to be upcast to object

jbrockmendel commented:
> Or are they generally a bit unreliable? 

Yes.  Usually ill run them all once, then re-run the ones that it claims are significantly changed and prune entries that aren't consistent-ish across runs

jbrockmendel commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -1845,6 +1845,14 @@ def test_constructor_series_copy(self, float_frame):
         assert not (series["A"] == 5).all()
 
+    def test_object_array_does_not_copy(self):
+        a = np.array(["a", "b"], dtype="object")
```
so if we have any columns that are inferred as datetimelike, we get copies of everything, i.e. dont retain any of the original array?

irgolic replied:
If one of the columns is inferred as datetimelike, each column is created as its own block. From my understanding, upon consolidation, these blocks get copied to a contiguous location.

jreback replied:
add the issue number as a comment

irgolic commented:
```
> asv continuous -f 1.1 9ed521e363959ea2aa2fd5c987aa08b5730a9454 309cf3a6c76c0671ac5710e14655257c9cbcdbf8 -b indexing.NumericSeriesIndexing.time_getitem_array   
...
       before           after         ratio
     [9ed521e3]       [309cf3a6]
-         331±3ms       12.5±0.1ms     0.04  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```
I could really only get this one to reproduce.

It seems that 309cf3a6 induced a performance boost. My PR undoes that performance boost.
```
> asv continuous -f 1.1 9ed521e363959ea2aa2fd5c987aa08b5730a9454 origin/2d-object-dont-copy -b indexing.NumericSeriesIndexing.time_getitem_array  
...
BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```
```
> asv continuous -f 1.1 309cf3a6c76c0671ac5710e14655257c9cbcdbf8 origin/2d-object-dont-copy -b indexing.NumericSeriesIndexing.time_getitem_array
...
       before           after         ratio
     [309cf3a6]       [7157e6aa]
+      12.5±0.2ms          332±3ms    26.69  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```
@jbrockmendel Any idea why?

irgolic commented:
I just needed to rebase. 
```
> asv continuous -f 1.1 upstream/master origin/2d-object-dont-copy -b indexing.NumericSeriesIndexing.time_getitem_array
...
BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```
We're good to go.

irgolic commented:
This PR is waiting for #38939, I assume.

simonjayhawkins commented:
> This PR is waiting for #38939, I assume.

@irgolic #38939 is merged. can you merge master to resolve conflicts

irgolic committed:
merge master

irgolic commented:
@simonjayhawkins merged, could you approve running lint and tests?

some changes have been made in dataframe constructors and block manager constructors since I wrote this. at a glance, dataframe constructor tests pass, and it should be fine, I'm rerunning asv though

simonjayhawkins commented:
> @simonjayhawkins merged, could you approve running lint and tests?

done

irgolic committed:
lint

irgolic commented:
@simonjayhawkins wow i think you might need to approve it every time I push

I'm too used to pushing and praying :x

irgolic committed:
Merge branch 'master' into 2d-object-dont-copy

irgolic committed:
add back _from_blocks

irgolic commented:
I wanted to remove `create_block_manager_from_blocks` as @jreback suggests here https://github.com/pandas-dev/pandas/pull/39272#discussion_r562652132

But then I found https://github.com/pandas-dev/pandas/pull/33892, where it says that dask uses it, so I left it in. Still, I added `create_block_manager_from_array`, which should be sufficient within the pandas codebase – I've removed all uses of `create_block_manager_from_blocks`.

irgolic commented:
Now this part of the test is failing:
```python
    def test_object_array_does_not_copy(self):
        a = np.array(["a", "b"], dtype="object")
        b = np.array([["a", "b"], ["c", "d"]], dtype="object")
        df = DataFrame(a)
>       assert np.shares_memory(df.values, a)
        df2 = DataFrame(b)
        assert np.shares_memory(df2.values, b)
```

It didn't fail before I merged master, and I can't reproduce this fail locally. I've tried on two machines, in an interpreter and with pytest.

Any idea why this happens in CI now? Might this be related to your change, @jbrockmendel?

irgolic committed:
try not replacing _from_blocks with _from_array

irgolic commented:
I tried fixing the CI fail by simplifying my solution (last commit), but it didn't change anything. Benchmarks are ok in both last and second-to-last commits, let me know which one you'd prefer.

jreback commented:
moving off the milestone, but can merge for 1.3 if fixed up

irgolic commented:
> moving off the milestone, but can merge for 1.3 if fixed up

@jreback, could you advise on how to proceed? Tests run fine locally, I can't reproduce the CI fail which appeared only recently.

jreback commented:
@irgolic you will likley need to create a local env just like the CI, use the .yaml files, alt you can just try to push and see what fixes things like this: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=60514&view=logs&j=69addb8c-d01a-5d3b-91bf-b74d46cb8150&t=cd1a349e-a8c9-5d71-a820-b09fec3303fd

irgolic commented:
@jreback a conda environment created from `environment.yml` did not reproduce it. I'll try bisecting and pushing, I suppose.

irgolic committed:
reduce test to only 2d arr

irgolic committed:
Merge branch 'master' into 2d-object-dont-copy

irgolic committed:
restore broader test

irgolic committed:
move whatsnew to v1.3.1

irgolic commented:
I've no idea why `np.shares_memory` was passing locally but failing in CI a month ago. Whatever it was, it seems like you guys fixed it.

The only tests failing on Azure now are 404 errors like the one failing master (`test_url`, `test_online_stylesheet`). Let me know if I'm missing anything @jreback, this looks good on my end.

jreback commented on doc/source/whatsnew/v1.3.1.rst:
```diff
@@ -24,7 +24,7 @@ Fixed regressions
 Bug fixes
 ~~~~~~~~~
--
+- Bug in :class:`DataFrame` constructor always copying 2D object arrays (:issue:`39272`)
```
move to 1.4

jreback commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -2080,6 +2080,14 @@ def test_constructor_series_copy(self, float_frame):
+        assert np.shares_memory(df.values, a)
+        df2 = DataFrame(b)
+        assert np.shares_memory(df2.values, b)
+
```
assert the frames are equal as well

irgolic replied:
did you mean assert the numpy arrays equal? They're not equal, but they do share memory:
```
>       np.testing.assert_array_equal(df.values, a)
E       AssertionError: 
E       Arrays are not equal
E       
E       (shapes (2, 1), (2,) mismatch)
E        x: array([['a'],
E                  ['b']], dtype=object)
E        y: array(['a', 'b'], dtype=object)
```
(note also I've split the test into two)

jreback commented on pandas/core/internals/construction.py:
```diff
@@ -351,14 +351,19 @@ def ndarray_to_mgr(
+                ]
+            else:
+                nb = new_block(values, placement=slice(len(columns)), ndim=2)
+                block_values = [nb]
```
can we remove the entire if/then here e.g. L350 and just use the inside logic (inlucding what you added), e.g. as L368 look duplicative now.

irgolic committed:
split test

irgolic committed:
Revert "try not replacing _from_blocks with _from_array"

This reverts commit 672f9c0be36cd6875e0570895b0d1b9230b4dfae.

irgolic committed:
move whatsnew to v1.4.0

irgolic committed:
annotate tests with link to issue

irgolic committed:
Merge branch 'master' into 2d-object-dont-copy

irgolic committed:
reword whatsnew

irgolic commented:
I finally figured out what was failing. It was the new ArrayManager. I just needed to add `@td.skip_array_manager_invalid_test` on the test.

The way the new array data management is set up currently is such that it copies all incoming numpy arrays. Our use case for pandas relies on 'wrapping' numpy arrays without copying them. Please consider this when enforcing the new array data manager. Although I'm assuming that this won't be standard for a while.

Until then, please accept this contribution which fixes the 2D object array copy bug in block manager. Note, I've also refactored that bit and made a `create_block_manager_from_array` methods instead of `_from_blocks`. `create_block_manager_from_blocks` is now not used anywhere in the pandas codebase anymore, but dask calls this pandas internal, so it should probably be kept there for now. 
@jreback Sorry for bouncing back and forth between the two implementations. I think this one's cleaner, let me know if I should change it back or amend it.

(also not sure why it says pandas-dev.pandas failed, all the tests passed)

jbrockmendel commented:
> The way the new array data management is set up currently is such that it copies all incoming numpy arrays. Our use case for pandas relies on 'wrapping' numpy arrays without copying them. Please consider this when enforcing the new array data manager. Although I'm assuming that this won't be standard for a while.

IIRC there was some discussion about this copy being avoided, not sure what happened to that.  Opening an issue about it might be worthwhile.

jbrockmendel commented on pandas/core/internals/construction.py:
```diff
@@ -342,37 +338,29 @@ def ndarray_to_mgr(
+            array = maybe_infer_to_datetimelike(array)
 
-    return create_block_manager_from_blocks(block_values, [columns, index])
+    return create_block_manager_from_array(array, [columns, index])
```
this change seems unnecessary.  id rather have this function return only here and not on L357

jbrockmendel commented on pandas/core/internals/construction.py:
```diff
@@ -342,37 +338,29 @@ def ndarray_to_mgr(
+            # don't convert (and copy) the objects if no type inference occurs
+            if any(
+                not is_dtype_equal(instance.dtype, array.dtype)
+                for instance in maybe_datetime
```
double-check me on this, but i think a more performant alternative would be

```
obj_columns = [x for x in array]
maybe_datetime = [maybe_infer_to_datetimelike(x) for x in obj_columns]
if any(x is not y for x, y in zip(obj_columns, maybe_datetime):
    [...]
```

irgolic replied:
I'm running asv to see if there's a performance difference

irgolic replied:
Here's the full asv run comparing the last commit (your suggestion) and the second to last commit. Any of these in particular you'd like me to rerun?
<details>
```
       before           after         ratio                                                                                                                  
     [dc2ae20d]       [96dd1b9e]                                                                                                                     
     <2d-object-dont-copy>       <2d-object-dont-copy~1>                                                                                                     
+     4.43±0.01ms       94.2±0.3ms    21.27  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Float64Index'>, 'uniqu
e_monotonic_inc')                                                                                                                                            
+        59.0±3ms       75.8±0.4ms     1.28  inference.ToTimedeltaErrors.time_convert('coerce')                                                              
+        45.8±5ms         58.5±5ms     1.28  gil.ParallelGroupbyMethods.time_loop(2, 'last')                                                                 
+      4.37±0.2ms       5.34±0.3ms     1.22  rolling.Engine.time_rolling_apply('Series', 'float', <function sum>, 'cython', 'max')                           
+      5.31±0.1ms       6.45±0.5ms     1.21  rolling.Engine.time_rolling_apply('DataFrame', 'float', <function Engine.<lambda>>, 'cython', 'sum')            
+      4.46±0.1ms       5.40±0.3ms     1.21  rolling.Engine.time_rolling_apply('DataFrame', 'float', <function sum>, 'cython', 'mean')                       
+      5.38±0.2ms       6.51±0.6ms     1.21  rolling.Engine.time_rolling_apply('DataFrame', 'int', <function Engine.<lambda>>, 'cython', 'max')              
+         147±9ms          178±1ms     1.21  inference.ToDatetimeISO8601.time_iso8601_tz_spaceformat                                                         
+        72.1±5ms         86.7±1ms     1.20  inference.ToDatetimeCache.time_dup_string_tzoffset_dates(False)                                                 
+     4.57±0.09ms       5.47±0.3ms     1.20  rolling.Engine.time_rolling_apply('DataFrame', 'float', <function sum>, 'cython', 'max')                        
+      5.23±0.2ms       6.23±0.4ms     1.19  rolling.Engine.time_rolling_apply('Series', 'float', <function Engine.<lambda>>, 'cython', 'mean')              
+      5.40±0.1ms       6.42±0.5ms     1.19  rolling.Engine.time_rolling_apply('DataFrame', 'float', <function Engine.<lambda>>, 'cython', 'median')         
+      4.29±0.1ms       5.10±0.3ms     1.19  rolling.Engine.time_rolling_apply('Series', 'int', <function sum>, 'cython', 'mean')                            
+      4.52±0.2ms       5.37±0.2ms     1.19  rolling.Engine.time_rolling_apply('DataFrame', 'int', <function sum>, 'cython', 'median')                       
+      4.45±0.2ms       5.28±0.4ms     1.19  rolling.Engine.time_rolling_apply('Series', 'float', <function sum>, 'cython', 'mean')                          
+      15.4±0.6μs       18.2±0.3μs     1.19  series_methods.NanOps.time_func('sum', 1000, 'boolean')                                                         
+      4.39±0.1ms       5.20±0.2ms     1.19  rolling.Engine.time_rolling_apply('Series', 'int', <function sum>, 'cython', 'max')                             
+      5.25±0.2ms       6.23±0.4ms     1.18  rolling.Engine.time_rolling_apply('Series', 'int', <function Engine.<lambda>>, 'cython', 'median')              
+      5.20±0.2ms       6.16±0.3ms     1.18  rolling.Engine.time_rolling_apply('Series', 'float', <function Engine.<lambda>>, 'cython', 'sum')               
+      4.37±0.2ms       5.17±0.3ms     1.18  rolling.Engine.time_rolling_apply('Series', 'int', <function sum>, 'cython', 'median')                          
+     4.53±0.04ms       5.35±0.3ms     1.18  rolling.Engine.time_rolling_apply('DataFrame', 'int', <function sum>, 'cython', 'min')                          
+      4.35±0.1ms       5.13±0.3ms     1.18  rolling.Engine.time_rolling_apply('Series', 'int', <function sum>, 'cython', 'sum')                             
+      4.40±0.1ms       5.18±0.3ms     1.18  rolling.Engine.time_rolling_apply('Series', 'float', <function sum>, 'cython', 'median')                        
+      5.25±0.2ms       6.17±0.2ms     1.17  rolling.Engine.time_rolling_apply('Series', 'float', <function Engine.<lambda>>, 'cython', 'min')               
+      5.39±0.2ms       6.32±0.4ms     1.17  rolling.Engine.time_rolling_apply('DataFrame', 'int', <function Engine.<lambda>>, 'cython', 'min')              
+        58.8±2ms         68.9±7ms     1.17  inference.ToTimedeltaErrors.time_convert('ignore')                                                              
+      4.42±0.2ms       5.17±0.2ms     1.17  rolling.Engine.time_rolling_apply('Series', 'float', <function sum>, 'cython', 'min')                           
+      5.33±0.1ms       6.23±0.2ms     1.17  rolling.Engine.time_rolling_apply('DataFrame', 'float', <function Engine.<lambda>>, 'cython', 'min')            
+      5.38±0.3ms       6.27±0.4ms     1.17  rolling.Engine.time_rolling_apply('Series', 'int', <function Engine.<lambda>>, 'cython', 'max')                 
+      4.73±0.3ms       5.51±0.4ms     1.17  rolling.Engine.time_expanding_apply('Series', 'int', <function sum>, 'cython', 'max')                           
+      5.20±0.1ms       6.06±0.3ms     1.16  rolling.Engine.time_rolling_apply('Series', 'int', <function Engine.<lambda>>, 'cython', 'min')                 
+      4.60±0.2ms       5.36±0.3ms     1.16  rolling.Engine.time_rolling_apply('DataFrame', 'int', <function sum>, 'cython', 'max')                          
+      13.2±0.4μs       15.3±0.3μs     1.16  series_methods.NanOps.time_func('max', 1000, 'boolean')                                                         
+     4.40±0.06ms       5.11±0.4ms     1.16  rolling.Engine.time_rolling_apply('Series', 'float', <function sum>, 'cython', 'sum')                           
+      4.57±0.1ms       5.29±0.3ms     1.16  rolling.Engine.time_rolling_apply('DataFrame', 'int', <function sum>, 'cython', 'sum')                          
+     4.67±0.09ms       5.41±0.2ms     1.16  rolling.Engine.time_rolling_apply('DataFrame', 'int', <function sum>, 'cython', 'mean')                         
+      5.39±0.1ms       6.23±0.5ms     1.16  rolling.Engine.time_rolling_apply('DataFrame', 'float', <function Engine.<lambda>>, 'cython', 'mean')           
+      5.21±0.1ms       6.02±0.4ms     1.16  rolling.Engine.time_rolling_apply('Series', 'float', <function Engine.<lambda>>, 'cython', 'max')               
+     4.59±0.03ms       5.29±0.2ms     1.15  rolling.Engine.time_rolling_apply('DataFrame', 'float', <function sum>, 'cython', 'median')                     
+      5.31±0.1ms       6.12±0.3ms     1.15  rolling.Engine.time_rolling_apply('DataFrame', 'int', <function Engine.<lambda>>, 'cython', 'mean')             
+     4.55±0.03ms       5.24±0.3ms     1.15  rolling.Engine.time_rolling_apply('DataFrame', 'float', <function sum>, 'cython', 'sum')                        
+      5.37±0.1ms       6.19±0.4ms     1.15  rolling.Engine.time_rolling_apply('DataFrame', 'int', <function Engine.<lambda>>, 'cython', 'sum')              
+      16.4±0.7μs         18.7±1μs     1.14  series_methods.SearchSorted.time_searchsorted('uint8')                                                          
+      5.26±0.1ms       5.99±0.3ms     1.14  rolling.Engine.time_rolling_apply('Series', 'float', <function Engine.<lambda>>, 'cython', 'median')            
+      4.64±0.1ms       5.29±0.3ms     1.14  rolling.Engine.time_rolling_apply('DataFrame', 'float', <function sum>, 'cython', 'min')                        
+     1.17±0.01ms      1.33±0.07ms     1.14  hash_functions.NumericSeriesIndexingShuffled.time_loc_slice(<class 'pandas.core.indexes.numeric.UInt64Index'>, 5
00000)                                                                                                                                   
+     2.58±0.03μs      2.93±0.08μs     1.13  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Float64Engine'>, <class 'numpy.f
loat64'>), 'monotonic_incr')                                                                                                                                 
+      5.21±0.2ms       5.91±0.3ms     1.13  rolling.Engine.time_rolling_apply('Series', 'int', <function Engine.<lambda>>, 'cython', 'mean')                
+        4.28±0ms       4.84±0.5ms     1.13  tslibs.tslib.TimeIntsToPydatetime.time_ints_to_pydatetime('time', 10000, datetime.timezone(datetime.timedelta(se
conds=3600)))                                                                                                                                                
+      4.51±0.2ms       5.10±0.2ms     1.13  rolling.Engine.time_rolling_apply('Series', 'int', <function sum>, 'cython', 'min')                             
+      5.50±0.2ms       6.21±0.2ms     1.13  rolling.Engine.time_rolling_apply('DataFrame', 'float', <function Engine.<lambda>>, 'cython', 'max')            
+      16.5±0.4μs         18.6±1μs     1.13  series_methods.SearchSorted.time_searchsorted('int16')                                                          
+     2.57±0.01μs       2.89±0.2μs     1.13  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Int64Engine'>, <class 'numpy.int
64'>), 'monotonic_incr')                                                                                                                                     
+       452±0.8ms         508±50ms     1.12  tslibs.tslib.TimeIntsToPydatetime.time_ints_to_pydatetime('time', 1000000, datetime.timezone(datetime.timedelta(
seconds=3600)))                                                                                                                                              
+        806±20μs         904±40μs     1.12  arithmetic.IntFrameWithScalar.time_frame_op_with_scalar(<class 'numpy.float64'>, 5.0, <built-in function ne>)   
+     3.20±0.04μs       3.59±0.2μs     1.12  tslibs.fields.TimeGetStartEndField.time_get_start_end_field(0, 'start', 'month', 'QS', 3)                       
+      13.2±0.5μs       14.8±0.6μs     1.12  series_methods.NanOps.time_func('min', 1000, 'boolean')                                                         
+         121±5μs          136±4μs     1.12  multiindex_object.GetLoc.time_large_get_loc                                                                     
+      47.6±0.2ms         53.0±1ms     1.11  rolling.TableMethod.time_apply('single')                                                                        
+     4.17±0.02μs       4.64±0.2μs     1.11  categoricals.CategoricalSlicing.time_getitem_scalar('non_monotonic')                                            
+      15.9±0.3ms       17.7±0.2ms     1.11  inference.ToTimedelta.time_convert_string_days                                                                  
+        741±20μs         824±20μs     1.11  arithmetic.IntFrameWithScalar.time_frame_op_with_scalar(<class 'numpy.int64'>, 2, <built-in function eq>)       
+        732±20μs         813±20μs     1.11  arithmetic.IntFrameWithScalar.time_frame_op_with_scalar(<class 'numpy.float64'>, 5.0, <built-in function le>)   
+     3.32±0.03μs       3.67±0.2μs     1.11  tslibs.fields.TimeGetStartEndField.time_get_start_end_field(0, 'start', 'year', 'B', 5)                         
+     3.44±0.02μs       3.81±0.3μs     1.11  tslibs.fields.TimeGetStartEndField.time_get_start_end_field(1, 'start', 'year', 'B', 5)                         
+       113±0.7μs          125±1μs     1.11  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'unique_monoton
ic_inc')                                                                                                                                                     
+     4.16±0.02μs       4.60±0.4μs     1.10  categoricals.CategoricalSlicing.time_getitem_scalar('monotonic_incr')                                           
+      13.0±0.2μs       14.3±0.2μs     1.10  series_methods.NanOps.time_func('sum', 1000, 'Int64')                                                           
+        60.3±1ms         66.6±1ms     1.10  rolling.Groupby.time_rolling_int('mean')                                                                        
+      14.4±0.1μs       15.9±0.5μs     1.10  series_methods.NanOps.time_func('min', 1000, 'Int64')                                                           
+        60.7±1ms         67.0±1ms     1.10  rolling.Groupby.time_rolling_int('max')                                                                         
+        61.5±1ms         67.9±1ms     1.10  rolling.Groupby.time_rolling_int('median')                                                                      
+        544±20ns         600±30ns     1.10  index_cached_properties.IndexCache.time_inferred_type('Int64Index')                                             
+     15.8±0.04μs       17.4±0.8μs     1.10  inference.MaybeConvertObjects.time_maybe_convert_objects                                                        
+       437±0.7ms         482±40ms     1.10  tslibs.tslib.TimeIntsToPydatetime.time_ints_to_pydatetime('time', 1000000, None)                                
+        60.5±2ms         66.6±1ms     1.10  rolling.Groupby.time_rolling_int('sum' (0))                                                                     
+     3.36±0.01μs       3.70±0.2μs     1.10  tslibs.fields.TimeGetStartEndField.time_get_start_end_field(0, 'end', 'year', 'B', 12)                          
+     3.24±0.04μs       3.57±0.2μs     1.10  tslibs.fields.TimeGetStartEndField.time_get_start_end_field(0, 'start', 'quarter', 'QS', 5)                     
+         142±2μs          157±1μs     1.10  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_mo
notonic_inc')                                                                                                                                                
+      41.1±0.3ms         45.2±1ms     1.10  frame_methods.Isnull.time_isnull_obj                                                                            
+     3.47±0.02μs       3.82±0.2μs     1.10  tslibs.fields.TimeGetStartEndField.time_get_start_end_field(1, 'end', 'quarter', 'QS', 12)                      
-      8.68±0.5μs      7.89±0.02μs     0.91  tslibs.resolution.TimeResolution.time_get_resolution('h', 0, datetime.timezone(datetime.timedelta(seconds=3600))
)                                                                                                                                                            
-      8.79±0.6μs      7.98±0.04μs     0.91  tslibs.resolution.TimeResolution.time_get_resolution('D', 1, datetime.timezone(datetime.timedelta(seconds=3600))
)                                                                                                                                                            
-        18.0±1μs      16.3±0.05μs     0.91  tslibs.resolution.TimeResolution.time_get_resolution('ns', 1, tzlocal())                                        
-      9.93±0.6μs      9.01±0.02μs     0.91  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(0, 4000, datetime.timezone(datetime.timedelta(sec
onds=3600)))                                                                                                                                                 
-      1.10±0.09s          998±5ms     0.91  arithmetic.OffsetArrayArithmetic.time_add_series_offset(<CustomBusinessMonthEnd> (0))
-      2.71±0.2μs         2.46±0μs     0.91  tslibs.tslib.TimeIntsToPydatetime.time_ints_to_pydatetime('date', 0, None)
-      9.98±0.7μs      9.05±0.03μs     0.91  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(0, 2000, datetime.timezone(datetime.timedelta(sec
onds=3600)))
-      2.75±0.2μs      2.50±0.02μs     0.91  tslibs.tslib.TimeIntsToPydatetime.time_ints_to_pydatetime('time', 0, tzlocal())
-      2.72±0.2μs         2.46±0μs     0.91  tslibs.tslib.TimeIntsToPydatetime.time_ints_to_pydatetime('timestamp', 0, None)
-      9.98±0.5μs      9.03±0.09μs     0.91  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(1, 5000, datetime.timezone(datetime.timedelta(sec
onds=3600)))

-      9.11±0.3μs       8.24±0.3μs     0.90  tslibs.timestamp.TimestampProperties.time_weekday_name(tzlocal(), None)
-      9.91±0.6μs       8.96±0.1μs     0.90  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(0, 12000, datetime.timezone(datetime.timedelta(se
conds=3600)))
-      10.0±0.6μs      9.06±0.04μs     0.90  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(1, 3000, datetime.timezone(datetime.timedelta(sec
onds=3600)))
-      10.1±0.8μs      9.09±0.03μs     0.90  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(0, 8000, datetime.timezone(datetime.timedelta(sec
onds=3600)))
-      9.94±0.7μs       8.99±0.1μs     0.90  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(1, 10000, datetime.timezone(datetime.timedelta(se
conds=3600)))
-        19.8±1μs      17.9±0.05μs     0.90  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(100, 2000, <DstTzInfo 'US/Pacific' LMT-1 day, 16:
07:00 STD>)
-      1.11±0.09s          998±3ms     0.90  arithmetic.OffsetArrayArithmetic.time_add_dti_offset(<CustomBusinessMonthEnd> (0))
-      9.91±0.6μs      8.95±0.08μs     0.90  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(0, 7000, datetime.timezone(datetime.timedelta(sec
onds=3600)))
-      9.99±0.5μs      9.01±0.09μs     0.90  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(1, 1000, datetime.timezone(datetime.timedelta(sec
onds=3600)))
-      2.58±0.2μs      2.32±0.01μs     0.90  tslibs.tslib.TimeIntsToPydatetime.time_ints_to_pydatetime('time', 0, datetime.timezone.utc)
-      9.97±0.7μs      8.96±0.06μs     0.90  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(1, 11000, datetime.timezone(datetime.timedelta(se
conds=3600)))
-      29.3±0.3ms       26.2±0.5ms     0.90  rolling.Apply.time_rolling('Series', 3, 'int', <built-in function sum>, False)
-      10.1±0.7μs      9.03±0.06μs     0.89  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(0, 9000, datetime.timezone(datetime.timedelta(sec
onds=3600)))
-      2.60±0.2μs      2.32±0.01μs     0.89  tslibs.tslib.TimeIntsToPydatetime.time_ints_to_pydatetime('timestamp', 0, datetime.timezone.utc)
-      10.0±0.7μs      8.98±0.04μs     0.89  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(1, 6000, datetime.timezone(datetime.timedelta(sec
onds=3600)))
-      8.65±0.2μs       7.73±0.1μs     0.89  tslibs.period.TimePeriodArrToDT64Arr.time_periodarray_to_dt64arr(1, 7000)
-      9.99±0.7μs      8.93±0.05μs     0.89  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(0, 2011, datetime.timezone(datetime.timedelta(sec
onds=3600)))
-        26.7±2μs       23.8±0.2μs     0.89  tslibs.timestamp.TimestampOps.time_normalize(tzlocal())
-      2.77±0.2μs      2.47±0.02μs     0.89  tslibs.tslib.TimeIntsToPydatetime.time_ints_to_pydatetime('time', 0, None)
-      6.29±0.6μs      5.60±0.08μs     0.89  tslibs.timedelta.TimedeltaConstructor.time_from_unit
-      11.6±0.8μs       10.3±0.1μs     0.89  tslibs.timestamp.TimestampOps.time_replace_None(datetime.timezone(datetime.timedelta(seconds=3600)))
-      10.1±0.6μs      8.95±0.06μs     0.89  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(0, 10000, datetime.timezone(datetime.timedelta(se
conds=3600)))
-      5.64±0.3μs      5.00±0.04μs     0.89  tslibs.timestamp.TimestampOps.time_normalize(None)
-        58.8±3μs       51.9±0.6μs     0.88  tslibs.timestamp.TimestampOps.time_normalize(tzfile('/usr/share/zoneinfo/Asia/Tokyo'))
-      3.19±0.3μs      2.80±0.02μs     0.88  tslibs.tslib.TimeIntsToPydatetime.time_ints_to_pydatetime('date', 1, None)
-        9.72±1μs      8.54±0.05μs     0.88  tslibs.tslib.TimeIntsToPydatetime.time_ints_to_pydatetime('timestamp', 0, datetime.timezone(datetime.timedelta(s
econds=3600)))
-      2.85±0.2μs      2.50±0.01μs     0.88  tslibs.tslib.TimeIntsToPydatetime.time_ints_to_pydatetime('timestamp', 0, tzlocal())
-     1.27±0.03ms      1.11±0.06ms     0.87  arithmetic.IntFrameWithScalar.time_frame_op_with_scalar(<class 'numpy.float64'>, 4, <built-in function mul>)
-        15.5±2μs       13.6±0.2μs     0.87  tslibs.timedelta.TimedeltaConstructor.time_from_components
-        8.74±1μs      7.57±0.05μs     0.87  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(1, 6000, <DstTzInfo 'US/Pacific' LMT-1 day, 16:07
:00 STD>)
-      8.05±0.8μs      6.92±0.03μs     0.86  tslibs.timestamp.TimestampOps.time_replace_None(<DstTzInfo 'US/Pacific' LMT-1 day, 16:07:00 STD>)
-     1.59±0.01ms      1.36±0.01ms     0.85  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(100, 4000, tzlocal())
-      6.61±0.6μs      5.65±0.08μs     0.85  tslibs.timedelta.TimedeltaConstructor.time_from_string
-        24.1±2μs       20.5±0.1μs     0.85  tslibs.timestamp.TimestampOps.time_normalize(datetime.timezone(datetime.timedelta(seconds=3600)))
-     11.7±0.05ms      9.84±0.06ms     0.84  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'nonunique_mon
otonic_inc')
-      1.66±0.1ms      1.39±0.07ms     0.84  series_methods.NanOps.time_func('prod', 1000000, 'int8')
-      10.8±0.2μs      9.05±0.04μs     0.84  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(0, 4006, datetime.timezone(datetime.timedelta(sec
onds=3600)))
-      10.7±0.3μs       8.95±0.2μs     0.83  tslibs.period.TimeDT64ArrToPeriodArr.time_dt64arr_to_periodarr(1, 1011, datetime.timezone(datetime.timedelta(sec
onds=3600)))
-      1.76±0.2ms       1.46±0.3ms     0.83  index_cached_properties.IndexCache.time_engine('MultiIndex')
-      57.4±0.4ms      34.8±0.07ms     0.61  algos.isin.IsinWithArange.time_isin(<class 'numpy.object_'>, 8000, 0)
-        93.1±1ms          400±3μs     0.00  hash_functions.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.Float64Index'>, 1000000)
```
</details>

irgolic replied:
Reran time_getitem_list_like, now it swung in the other direction, seems like it's just doing whatever it feels like.
```
       before           after         ratio
     [dc2ae20d]       [96dd1b9e]
     <2d-object-dont-copy>       <2d-object-dont-copy~1>
-     59.7±0.08ms      3.17±0.02ms     0.05  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'unique_monotonic_inc')
```

irgolic committed:
Revert "Revert "try not replacing _from_blocks with _from_array""

This reverts commit 9e00908665b8903598f836dfb8d326e6bfa8f38e.

irgolic committed:
Drop outside if/else

irgolic committed:
better performance?

irgolic commented:
This is what it looks like now:
```python
    if dtype is None and is_object_dtype(values.dtype):
        obj_columns = [x for x in values]
        maybe_datetime = [maybe_infer_to_datetimelike(x) for x in obj_columns]
        # don't convert (and copy) the objects if no type inference occurs
        if any(x is not y for x, y in zip(obj_columns, maybe_datetime)):
            dvals_list = [ensure_block_shape(dval, 2) for dval in maybe_datetime]
            block_values = [
                new_block(dvals_list[n], placement=n, ndim=2)
                for n in range(len(dvals_list))
            ]
        else:
            nb = new_block(values, placement=slice(len(columns)), ndim=2)
            block_values = [nb]
    else:
        nb = new_block(values, placement=slice(len(columns)), ndim=2)
        block_values = [nb]
```
Still not too pretty because of the double else. We could avoid it by abstracting into a separate private function. However without a walrus operator, we'd need to call `maybe_infer_to_datetimelike` twice (unless I'm missing something). Would that be fine, or should I leave it as is?

jbrockmendel commented:
> Still not too pretty because of the double else

im not bothered by this at all

irgolic committed:
lint

jreback commented:
thanks @irgolic thanks for sticking with it! very nice!
